### PR TITLE
[counters] Optimize PUT/PATCH /players/<id>/counters

### PR DESCRIPTION
**Pull request by judgeaxl**
- Batch generate counters, instead of running redis/query/create for each one
- Batch generate player counters, instead of running query/create for each one
- Batch update counter entries, instead of running insert for each set of periods
- Relies on custom postgresql features, ON CONFLICT and RETURNING
  to efficiently UPSERT multiple columns while returning the final
  values in a non-complex statement. Essentially, all queries now take the form of:
```
INSERT INTO table (col1, ..., colN) VALUES (v1a, ..., vNa), (v1P, ..., vNP)
ON CONFLICT (col1, ..., colN) DO UPDATE SET <expression>;
``` 
- Migrate PlayerCounter and CounterEntry primary keys to BigInt
- Ensure uniqueness of PlayerCounter and CounterEntry rows

### CODE REVIEW

File: driftbase/api/players/counters.py [hunk 1]
```diff
@@ -152,88 +156,62 @@ def _patch(self, player_id):
...
+                                             context_id=context_id,
+                                             is_absolute=is_absolute,
+                                             timestamp=timestamp
+                                             )
 
-        g.db.commit()
+        counter_ids = []
+        counters = batch_get_or_create_counters([(k, v["counter_type"]) for k, v in counter_updates.items()])
+        for (counter_id, name) in counters:
+            counter_updates[name].update(dict(counter_id=counter_id))
```
Comment (einarth) [inline driftbase/api/players/counters.py:205]: `counter_updates[name]["counter_id"] = counter_id` is clearer and faster.

File: driftbase/counters.py [hunk 2]
```diff
@@ -54,60 +54,71 @@ def get_counter(counter_key):
...
-        g.db.add(player_counter)
-    g.db.commit()
-    return player_counter
+    if len(counter_values):
+        insert_clause = insert(CounterEntry).values(counter_values)
+        update_clause = insert_clause.on_conflict_do_update(
+            index_elements=['counter_id', 'player_id', 'period', 'date_time'],
+            set_=dict(value=CounterEntry.value + insert_clause.excluded.value))
+        db_session.execute(update_clause)
+        db_session.commit()
```
Comment (einarth) [inline driftbase/counters.py:121]: Is there a reason we commit() inside both if-blocks rather than once at the end?

**Review by einarth — APPROVED** [review]

File: driftbase/api/players/counters.py (hunk 1 above)
Comment (judgeaxl) [reply driftbase/api/players/counters.py:205]: Right, I was updating multiple properties inititally

File: driftbase/counters.py (hunk 2 above)
Comment (judgeaxl) [reply driftbase/counters.py:121]: Not anymore, there was an issue with the query when it had a different format, but it should be fined to commit both at once

Comment (judgeaxl) [reply driftbase/counters.py]: Potentially there's a scenario where we'd want to split the update into multiple statements due to size, but I need to find if there's such a limit first

**Pull request merged**